lib/web_ui/lib/src/engine/keyboard_binding.dart

```diff
@@ -0,0 +1,489 @@
+    final int logicalKey = () {
+      if (kWebLogicalLocationMap.containsKey(event.key!)) {
+        final int? result = kWebLogicalLocationMap[event.key!]?[event.location!];
+        assert(result != null, 'Invalid modifier location: ${event.key}, ${event.location}');
+        return result!;
+      }
+      if (character != null)
+        return _characterToLogicalKey(character);
+      if (eventKey == _kLogicalDead)
+        return _deadKeyToLogicalKey(physicalKey, event);
+      return _otherLogicalKey(eventKey);
+    }();
+
+    assert(event.type == 'keydown' || event.type == 'keyup');
+    final bool isPhysicalDown = event.type == 'keydown' ||
+      // On macOS, both keydown and keyup events of CapsLock should be considered keydown,
+      // followed by an immediate cancel event.
+      (_shouldSynthesizeCapsLockUp() && event.code! == _kPhysicalCapsLock);
+
+    final int? lastLogicalRecord = _pressingRecords[physicalKey];
+
+    ui.KeyChange change;
+
+    if (_shouldSynthesizeCapsLockUp() && event.code! == _kPhysicalCapsLock) {
+      // Case 1: Handle CapsLock on macOS
+      //
+      // On macOS, both keydown and keyup events of CapsLock are considered
+      // keydown, followed by an immediate synchronized up event.
+
+      _scheduleAsyncEvent(
+        Duration.zero,
+        () => ui.KeyData(
+          timeStamp: timeStamp,
+          change: ui.KeyChange.up,
+          physical: physicalKey,
+          logical: logicalKey,
+          character: null,
+          synthesized: true,
+        ),
+        () {
+          _pressingRecords.remove(physicalKey);
+        }
+      );
+      change = ui.KeyChange.down;
+
+    } else if (isPhysicalDown) {
+      // Case 2: Handle key down of normal keys
+      change = ui.KeyChange.down;
+      if (lastLogicalRecord != null) {
+        // This physical key is being pressed according to the record.
+        if (event.repeat ?? false) {
+          // A normal repeated key.
+          change = ui.KeyChange.repeat;
+        } else {
+          // A non-repeated key has been pressed that has the exact physical key as
+          // a currently pressed one, usually indicating multiple keyboards are
+          // pressing keys with the same physical key, or the up event was lost
+          // during a loss of focus. The down event is ignored.
+          return false;
+        }
+      } else {
+        // This physical key is not being pressed according to the record. It's a
+        // normal down event, whether the system event is a repeat or not.
+      }
+
+    } else { // isPhysicalDown is false and not CapsLock
+      // Case 2: Handle key up of normal keys
+      if (lastLogicalRecord == null) {
+        // The physical key has been released before. It indicates multiple
+        // keyboards pressed keys with the same physical key. Ignore the up event.
+        return false;
+      }
+
+      change = ui.KeyChange.up;
+    }
+
+    late final int? nextLogicalRecord;
```

Comment:
You can drop the `late` here, as long as the switch statement below is exhaustive and the static analyzer can verify that `nextLogicalRecord` is always initialized exactly one time.